### DISCUSSION TITLE
TDD authenatication w/ magic links

### DISCUSSION THREAD

**Discussion by dompatmore** [supabase/supabase#6367] [Questions]
Hello,
   I like to take a TDD approach in my development work, and I'm struggling with how to how to test a workflow with magic link authentication.  My current thought is when switching the environment to test, so username/password authenatication, but I'm wonder if there's better strategies/best practice available. Thoughts?

**Comment by egor-romanov** [comment]
Hi!

You can find some info regarding magic link testing in [that issue](https://github.com/supabase/supabase/discussions/4381).

As for extracting magic link from email in your test, you may find a lot information about the tools and libraries you can use on the internet, but this would be specific for your tech stack. Once you parse email letter you may find a 'magic link' and use it for login.

I think if I would do such kind of testing, I would create a test suite for different login types (all enabled types for your app) to just check directly if you set up your environments right and that your email templates for example won't change accidentally. That would test your login flow directly.
And for all other tests I would set up the easiest and fastest login hook using login/pass. Maybe even signing in once per all tests or per suite.
